### PR TITLE
Policy provision

### PR DESCRIPTION
**Pull request by carleton-dds**
https://www.pivotaltracker.com/n/projects/2160940/stories/170267266

Completed with @graham-dds

Adds policy set to root management group during provisioning process.

A [policy set](https://docs.microsoft.com/en-us/azure/governance/policy/overview#initiative-definition) is a grouping of two or more policy definitions that can be assigned at once. The 3 policy definitions included with this PR concern:

1. Region restrictions
2. VM SKU (image type) restrictions
3. Resource restrictions

While assembling this PR we included some code that could be improved in a later PR. Specifically, [each of the REST API calls](https://github.com/dod-ccpo/atst/blob/policy-provision/atat/domain/csp/cloud/azure_cloud_provider.py#L279-L304) we make to Azure reuse nearly identical Try/Catch blocks to handle any potential errors in the API response. It might be possible to abstract this into a reusable method to make the code a little DRY-er

### CODE REVIEW

File: alembic/versions/137f4e27aa71_add_policies_to_portfolio_state_machine.py [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+            "BILLING_OWNER_IN_PROGRESS",
+            "BILLING_OWNER_FAILED",
+            "POLICIES_CREATED",
+            "POLICIES_IN_PROGRESS",
+            "POLICIES_FAILED",
+            name="fsmstates",
+            native_enum=False,
+                ),
+        existing_nullable=False,
+    )
```
Comment (dandds) [inline alembic/versions/137f4e27aa71_add_policies_to_portfolio_state_machine.py]: Woof, this looks more cumbersome every time. We might want to add a story to address this; there's gotta be a way to enforce the enum as a constraint on the column without having to do this every time we make a change.

File: atat/domain/csp/cloud/azure_cloud_provider.py [hunk 2]
```diff
@@ -271,48 +273,154 @@ def _get_management_group(self, credentials, management_group_id):
...
+                    "policyDefinitionReferenceId": definition["properties"][
+                        "displayName"
+                    ],
+                    "parameters": policy.parameters,
+                }
+            )
+        policy_set_definition = self._create_policy_set(
+            policy_session,
+            payload.root_management_group_id,
+            "Default JEDI Policy Set",
```
Comment (dandds) [inline atat/domain/csp/cloud/azure_cloud_provider.py]: Can we make this a module constant?

File: policies/portfolios/vm-sku-restrictions.json [hunk 3]
```diff
@@ -0,0 +1,157 @@
+{
+  "definitionPoint": "portfolio",
+  "policyDefinition": {
+    "properties": {
+      "displayName": "SKUs",
```
Comment (dandds) [inline policies/portfolios/vm-sku-restrictions.json]: Can we put something more specific here? "Allowed SKUs" or something like that?

File: tests/domain/cloud/test_azure_csp.py [hunk 4]
```diff
@@ -1663,3 +1639,41 @@ def _validate_domain_name(mock_azure, name):
             _validate_domain_name,
         )
         assert mock_azure.generate_valid_domain_name(tenant_name) != tenant_name
+
+
+def test_create_policies(mock_azure: AzureCloudProvider, monkeypatch):
+    final_assignment_id = "whatever"
+    post_results = [
```
Comment (dandds) [inline tests/domain/cloud/test_azure_csp.py]: Quibble, but `put_results` maybe? 😁

File: atat/domain/csp/cloud/azure_cloud_provider.py [hunk 5]
```diff
@@ -271,48 +273,154 @@ def _get_management_group(self, credentials, management_group_id):
         response = mgmgt_group_client.management_groups.get(management_group_id)
         return response
 
-    def _create_policy_definition(
-        self, credentials, subscription_id, management_group_id, properties,
+    def _create_policy_definition(self, session, root_management_group_id, policy):
+        create_policy_definition_uri = f"{self.sdk.cloud.endpoints.resource_manager}providers/Microsoft.Management/managementgroups/{root_management_group_id}/providers/Microsoft.Authorization/policyDefinitions/{policy.definition['properties']['displayName']}?api-version=2019-09-01"
+        body = policy.definition
+        try:
+            result = session.put(create_policy_definition_uri, json=body, timeout=30,)
```
[the remark this answers is not in the document]
Comment (carleton-dds) [reply atat/domain/csp/cloud/azure_cloud_provider.py]: As far as we can tell everything seems to be idempotent, using the list policy REST route returns a consistent number of runs of the sample script.

create-new-policy-set-definition returns 200 on repeat runs after the initial creation, but all other end points consistently return 201. (Thanks Azure!) I'll post a small update to make this clearer

File: policies/portfolios/vm-sku-restrictions.json (hunk 3 above)
Comment (carleton-dds) [reply policies/portfolios/vm-sku-restrictions.json]: Done! https://github.com/dod-ccpo/atst/pull/1459/commits/26351c5c0300c292ab417fd57f10e5ce8e1619bc

File: tests/domain/cloud/test_azure_csp.py (hunk 4 above)
Comment (carleton-dds) [reply tests/domain/cloud/test_azure_csp.py]: Done! https://github.com/dod-ccpo/atst/pull/1459/commits/e158e73051ee1e07b324d412bb221a2e7d8ddf97

File: atat/domain/csp/cloud/azure_cloud_provider.py [hunk 6]
```diff
@@ -271,48 +273,154 @@ def _get_management_group(self, credentials, management_group_id):
         response = mgmgt_group_client.management_groups.get(management_group_id)
         return response
 
-    def _create_policy_definition(
-        self, credentials, subscription_id, management_group_id, properties,
+    def _create_policy_definition(self, session, root_management_group_id, policy):
```
[the remark this answers is not in the document]
Comment (carleton-dds) [reply atat/domain/csp/cloud/azure_cloud_provider.py]: I'm not sure I have enough context to answer the question of whether this should be a generic internal method, but I'll be happy to change the parameter and make it a public method if you think that's best

[the remark this answers is not in the document]
Comment (dandds) [reply atat/domain/csp/cloud/azure_cloud_provider.py]: Let's leave it for now. Making it generic might be over-planning at this stage.

File: atat/domain/csp/cloud/azure_cloud_provider.py (hunk 2 above)
Comment (carleton-dds) [reply atat/domain/csp/cloud/azure_cloud_provider.py]: Done! https://github.com/dod-ccpo/atst/pull/1459/commits/75ca2f63191e794c1e2f75fb232d9a55541779e1

File: atat/domain/csp/cloud/azure_cloud_provider.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (carleton-dds) [reply atat/domain/csp/cloud/azure_cloud_provider.py]: https://github.com/dod-ccpo/atst/pull/1459/commits/c9cb3d5c17c6c9c8bb61ba2b72fd560b10e9b40d

File: alembic/versions/137f4e27aa71_add_policies_to_portfolio_state_machine.py (hunk 1 above)
Comment (carleton-dds) [reply alembic/versions/137f4e27aa71_add_policies_to_portfolio_state_machine.py]: https://www.pivotaltracker.com/story/show/171734708
At the top of the ice box. Not sure where's the best place to put new stories like this

**Comment by carleton-dds** [comment]
Ready for Re-review

**Pull request merged**